1. Read your Titanic dataset as usual: Create a training set and a testing set then apply decision tree. 

loading dataset

In [1]:
import pandas as pd

df = pd.read_csv(r'C:\Users\Meriem\Desktop\AI\titanic-passengers.csv', sep=";")

print(df.head().to_string(),'\n \n') 
df.info()

   PassengerId Survived  Pclass                                         Name     Sex   Age  SibSp  Parch      Ticket     Fare  Cabin Embarked
0          343       No       2                   Collander, Mr. Erik Gustaf    male  28.0      0      0      248740  13.0000    NaN        S
1           76       No       3                      Moen, Mr. Sigurd Hansen    male  25.0      0      0      348123   7.6500  F G73        S
2          641       No       3                       Jensen, Mr. Hans Peder    male  20.0      0      0      350050   7.8542    NaN        S
3          568       No       3  Palsson, Mrs. Nils (Alma Cornelia Berglund)  female  29.0      0      4      349909  21.0750    NaN        S
4          672       No       1                       Davidson, Mr. Thornton    male  31.0      1      0  F.C. 12750  52.0000    B71        S 
 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype

preprocessing dataset 

In [2]:
def preprocess_data(new_data):
    new_data['Age'].fillna(new_data['Age'].mean(),inplace=True)
    new_data.replace({'Sex':{'male': 1,'female':0}},inplace=True)
    new_data['Cabin']=new_data.Cabin.fillna('G6')
    new_data.replace({'Survived':{'Yes': 1,'No':0}},inplace=True)
    return new_data
data=preprocess_data(df)

splitting dataset into training and testing set and applying tree algorithm

In [40]:
from sklearn.model_selection import train_test_split
from sklearn import tree   
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
import graphviz 
from sklearn.tree import export_graphviz
import matplotlib.pyplot as plt
from sklearn.tree import plot_tree
from sklearn import datasets
from sklearn.ensemble import RandomForestClassifier 
from sklearn import metrics  
from sklearn.linear_model import LogisticRegression

In [5]:

x=data.drop(["Survived", "Name", "Cabin", "Ticket", "Embarked"], axis=1)
y= data["Survived"]


x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.20,random_state=10)

clf = tree.DecisionTreeClassifier()  
clf.fit(x_train, y_train)   
y_pred=clf.predict(x_test)   
print("score:{}".format(accuracy_score(y_test, y_pred)))

score:0.7877094972067039


2. Plot your decision tree and try to read the tree branches and conclude a prediction manually.

In [17]:
dot_data = tree.export_graphviz(clf, out_file=None)
graph = graphviz.Source(dot_data)
graph.render(view=True)
text_representation = tree.export_text(clf)
print(text_representation)

|--- feature_2 <= 0.50
|   |--- feature_1 <= 2.50
|   |   |--- feature_3 <= 2.50
|   |   |   |--- class: 0
|   |   |--- feature_3 >  2.50
|   |   |   |--- feature_6 <= 26.12
|   |   |   |   |--- feature_0 <= 48.00
|   |   |   |   |   |--- class: 0
|   |   |   |   |--- feature_0 >  48.00
|   |   |   |   |   |--- feature_3 <= 55.50
|   |   |   |   |   |   |--- feature_4 <= 0.50
|   |   |   |   |   |   |   |--- feature_0 <= 206.00
|   |   |   |   |   |   |   |   |--- feature_0 <= 195.50
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |--- feature_0 >  195.50
|   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- feature_0 >  206.00
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- feature_4 >  0.50
|   |   |   |   |   |   |   |--- feature_5 <= 0.50
|   |   |   |   |   |   |   |   |--- feature_3 <= 43.00
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |--- feature_3 >  4

there are 712 data records in the node, the data is distributed to 444, and 268.
After the first split based on X[2] <= 0.5, all samples meeting the criteria are placed on the left, which are 252. Those who don't (460) are put into the node on the right.

3. Change the decision tree parameters(at least two parameters).  pruning

In [18]:
dtree = tree.DecisionTreeClassifier(criterion = "gini", splitter = "random", max_leaf_nodes = 10, min_samples_leaf = 5, max_depth = 5)

4. Calculate the new accuracy and compare it with the previous results. 

In [19]:
dtree.fit(x_train, y_train)   
y_pred=dtree.predict(x_test)   
print("score:{}".format(accuracy_score(y_test, y_pred)))

score:0.770949720670391


In [20]:
dot_data = tree.export_graphviz(dtree, out_file=None)
graph = graphviz.Source(dot_data)
graph.render(view=True)
text_representation = tree.export_text(dtree)
print(text_representation)

|--- feature_2 <= 0.72
|   |--- feature_1 <= 2.85
|   |   |--- class: 1
|   |--- feature_1 >  2.85
|   |   |--- feature_5 <= 3.93
|   |   |   |--- feature_5 <= 0.41
|   |   |   |   |--- class: 1
|   |   |   |--- feature_5 >  0.41
|   |   |   |   |--- feature_6 <= 32.67
|   |   |   |   |   |--- class: 0
|   |   |   |   |--- feature_6 >  32.67
|   |   |   |   |   |--- class: 0
|   |   |--- feature_5 >  3.93
|   |   |   |--- class: 0
|--- feature_2 >  0.72
|   |--- feature_1 <= 2.58
|   |   |--- feature_1 <= 1.66
|   |   |   |--- feature_3 <= 45.49
|   |   |   |   |--- class: 0
|   |   |   |--- feature_3 >  45.49
|   |   |   |   |--- class: 0
|   |   |--- feature_1 >  1.66
|   |   |   |--- feature_5 <= 0.49
|   |   |   |   |--- class: 0
|   |   |   |--- feature_5 >  0.49
|   |   |   |   |--- class: 0
|   |--- feature_1 >  2.58
|   |   |--- class: 0



the accuracy after the pruning is slightly lower than the decision tree (0.770949720670391< 0.7877094972067039). the decision tree is more accurate.

5. Use a random forest then change the number of estimators.

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=44)
rf_model = RandomForestClassifier(n_estimators=50)
rf_model.fit(X_train, y_train)

RandomForestClassifier(n_estimators=50)

6. Calculate the new accuracy and compare it with the previous result.

In [44]:
predictions = rf_model.predict(X_test)
print("score:{}".format(accuracy_score(y_test, predictions)))

score:0.9566666666666667


the accuracy of the random forest model is higher than the previous (0.9566666666666667>0.770949720670391), and also higher than the decision tree so the accuracy of the random forest model is the best.